File: packages/i18n/src/format.ts
/**
 * The locale-aware primitives, shared; the presentation is not.
 *
 * `packages/web` renders a token budget as `8,192` and the CLI renders the same
 * number as `1.2k`, and that difference is deliberate — a settings panel has a
 * column to spend and a terminal status line does not. So what is shared here is
 * the part that is *locale* knowledge (which separator, which unit word, which
 * plural category) and not the part that is *layout* judgement.
 *
 * Every function takes the locale explicitly. That is the entire fix for the
 * bug the old `formatTokens` was working around: it hand-rolled comma grouping
 * because `toLocaleString()` on a machine set to `de-DE` renders 8192 as
 * `8.192`, which reads as eight in the one panel whose job is making a budget
 * legible. The problem was never that grouping is locale-aware — it is that the
 * locale was *implicit*, and so was whatever the machine happened to be set to.
 * Passed in, it is the install's locale on every machine.
 *
 * The `Intl` constructors are memoised because they are expensive to build and
 * are called per row of a transcript.
 */

/** A number with the locale's grouping separator. */
export function formatNumber(value: number, locale: string): string {
  if (!Number.isFinite(value)) return '—';
  return numberFormat(locale).format(value);
}

/** A number shortened to three significant digits — `1.2K`, `8.2K`, `1.1M`. */
export function formatCompactNumber(value: number, locale: string): string {
  if (!Number.isFinite(value)) return '—';
  return compactNumberFormat(locale).format(value);
}

/**
 * An absolute date, for anything too old to phrase as an interval.
 *
 * `timeZone` is optional and omitting it means the host zone, which is what
 * every `Intl` constructor already does — so a caller that has no opinion is
 * unchanged. A caller that does have one passes the install's `ui.timezone`,
 * and the point of threading it this far is the same as threading the locale:
 * an implicit zone is whatever the machine happens to be set to, and two
 * machines then disagree about a date that is a single instant.
 */
export function formatDate(
  atMs: number,
  locale: string,
  timeZone?: string,
): string {
  if (!Number.isFinite(atMs)) return '—';
  return dateFormat(locale, timeZone).format(new Date(atMs));
}

/**
 * A date *and* a time, with the zone named.
 *
 * Separate from `formatDate` rather than an option on it, because the two
 * answer different questions and the wrong one is silently wrong: a file's
 * modified date reads better without a time, and a scheduled job's next run is
 * *only* about the time — rendered through `formatDate` it reads `8 Aug 2026`
 * and drops the one field it exists for.
 *
 * `timeZoneName: 'short'` is not decoration. Once an install can render in a
 * zone that is not the reader's own, an unlabelled `09:00` is a number the
 * reader will assume is their clock. The label is what makes it checkable.
 */
export function formatDateTime(
  atMs: number,
  locale: string,
  timeZone?: string,
): string {
  if (!Number.isFinite(atMs)) return '—';
  return dateTimeFormat(locale, timeZone).format(new Date(atMs));
}

/**
 * Which interval an instant falls into, without saying it in any language.
 *
 * Split from the wording because the two halves belong to different owners. The
 * *thresholds* are shared — a minute is a minute in every locale, and the rules
 * about when to stop counting are product decisions rather than language ones.
 * The *wording* is not: the sidebar says `just now` where the bare `Intl` output
 * says `now`, and that is a copy decision the web app is entitled to keep.
 *
 * `now` is a parameter rather than a `Date.now()` call so the boundaries can be
 * tested without a fake clock, and so a list re-rendering mid-scroll cannot show
 * two rows measured against two different instants.
 */
export type RelativeSpan =
  /** Inside the last minute. The caller supplies its own wording. */
  | { readonly kind: 'now' }
  /** Old enough that an interval reads worse than a date. */
  | { readonly kind: 'date' }
  /** Not a timestamp at all. */
  | { readonly kind: 'unknown' }
  | {
      readonly kind: 'ago';
      readonly value: number;
      readonly unit: Intl.RelativeTimeFormatUnit;
    };

export function relativeSpan(atMs: number, now: number): RelativeSpan {
  if (!Number.isFinite(atMs)) return { kind: 'unknown' };

  const elapsed = now - atMs;

  // A timestamp slightly in the *future* is the present rather than a negative
  // interval: the server and the browser have separate clocks, and a few
  // seconds of skew is normal rather than an error worth rendering.
  if (elapsed < 60_000) return { kind: 'now' };

  const minutes = Math.floor(elapsed / 60_000);
  if (minutes < 60) return { kind: 'ago', value: minutes, unit: 'minute' };

  const hours = Math.floor(minutes / 60);
  if (hours < 24) return { kind: 'ago', value: hours, unit: 'hour' };

  const days = Math.floor(hours / 24);
  // Past a week the relative form stops being informative — "23d ago" is worse
  // than a date, because nobody counts back three weeks in their head.
  return days < 7
    ? { kind: 'ago', value: days, unit: 'day' }
    : { kind: 'date' };
}

/**
 * An interval, worded.
 *
 * `narrow` by default because that is what both surfaces want — `5m ago` in a
 * sidebar row and in a terminal line alike. `long` is there for anywhere with
 * room to spell it out.
 */
export function formatRelativeSpan(
  span: RelativeSpan,
  locale: string,
  style: Intl.RelativeTimeFormatStyle = 'narrow',
): string | undefined {
  if (span.kind !== 'ago') return undefined;
  return relativeTimeFormat(locale, style).format(-span.value, span.unit);
}

/**
 * A duration broken into its parts, for a caller to word.
 *
 * Returns numbers rather than a string because the two surfaces disagree about
 * the wording and agree about the arithmetic — and the arithmetic is the half
 * with the boundaries that are easy to get wrong and invisible when they are:
 * 59.9 seconds rounding to "60s", a sub-millisecond tool call reporting "0ms"
 * as though it never ran.
 */
interface DurationParts {
  readonly unit: 'ms' | 'second' | 'minute' | 'hour';
  /** The leading number, already rounded for display. */
  readonly value: number;
  /** The remainder in the next unit down, for `2m 30s`. Zero above an hour. */
  readonly remainder: number;
  /** True below ten seconds, where one decimal is worth showing. */
  readonly fractional: boolean;
}

export function durationParts(ms: number): DurationParts | undefined {
  if (!Number.isFinite(ms) || ms < 0) return undefined;
  if (ms < 1000) {
    return {
      unit: 'ms',
      value: Math.round(ms),
      remainder: 0,
      fractional: false,
    };
  }

  const totalSeconds = Math.floor(ms / 1000);
  if (totalSeconds < 60) {
    // One decimal below ten seconds, where the difference is worth seeing.
    const fractional = totalSeconds < 10;
    return {
      unit: 'second',
      value: fractional ? Math.round((ms / 1000) * 10) / 10 : totalSeconds,
      remainder: 0,
      fractional,
    };
  }

  const minutes = Math.floor(totalSeconds / 60);
  if (minutes < 60) {
    return {
      unit: 'minute',
      value: minutes,
      remainder: totalSeconds % 60,
      fractional: false,
    };
  }

  const hours = Math.floor(minutes / 60);
  return {
    unit: 'hour',
    value: hours,
    remainder: minutes % 60,
    fractional: false,
  };
}

/**
 * Which CLDR plural category a count falls into for this locale.
 *
 * i18next resolves plurals through the same `Intl.PluralRules`, so this exists
 * for the places that build a string outside a resource — not as a second
 * pluralisation scheme.
 */
export function pluralCategory(
  count: number,
  locale: string,
): Intl.LDMLPluralRule {
  return pluralRules(locale).select(count);
}

// Memoised constructors

const numberFormats = new Map<string, Intl.NumberFormat>();
const compactNumberFormats = new Map<string, Intl.NumberFormat>();
const dateFormats = new Map<string, Intl.DateTimeFormat>();
const dateTimeFormats = new Map<string, Intl.DateTimeFormat>();
const relativeTimeFormats = new Map<string, Intl.RelativeTimeFormat>();
const pluralRuleSets = new Map<string, Intl.PluralRules>();

function numberFormat(locale: string): Intl.NumberFormat {
  return cached(numberFormats, locale, () => new Intl.NumberFormat(locale));
}

function compactNumberFormat(locale: string): Intl.NumberFormat {
  return cached(
    compactNumberFormats,
    locale,
    () =>
      new Intl.NumberFormat(locale, {
        notation: 'compact',
        compactDisplay: 'short',
        maximumSignificantDigits: 2,
      }),
  );
}

/**
 * The zone as part of the memo key, not just as an option.
 *
 * Keyed on both because the cache is keyed on everything the constructor was
 * given — a map keyed on the locale alone would hand a caller asking for Tokyo
 * the formatter built earlier for UTC, and the bug would be invisible until
 * someone compared two rows.
 */
function formatterKey(locale: string, timeZone: string | undefined): string {
  return timeZone === undefined ? locale : `${locale}/${timeZone}`;
}

/** `{ timeZone }` only when there is one — passing `undefined` is not the same as omitting it. */
function zoneOption(timeZone: string | undefined): { timeZone?: string } {
  return timeZone === undefined ? {} : { timeZone };
}

function dateFormat(locale: string, timeZone?: string): Intl.DateTimeFormat {
  return cached(
    dateFormats,
    formatterKey(locale, timeZone),
    () =>
      new Intl.DateTimeFormat(locale, {
        ...zoneOption(timeZone),
        year: 'numeric',
        month: 'short',
        day: 'numeric',
      }),
  );
}

function dateTimeFormat(
  locale: string,
  timeZone?: string,
): Intl.DateTimeFormat {
  return cached(
    dateTimeFormats,
    formatterKey(locale, timeZone),
    () =>
      new Intl.DateTimeFormat(locale, {
        ...zoneOption(timeZone),
        year: 'numeric',
        month: 'short',
        day: 'numeric',
        hour: '2-digit',
        minute: '2-digit',
        timeZoneName: 'short',
      }),
  );
}

function relativeTimeFormat(
  locale: string,
  style: Intl.RelativeTimeFormatStyle,
): Intl.RelativeTimeFormat {
  return cached(
    relativeTimeFormats,
    `${locale}/${style}`,
    () => new Intl.RelativeTimeFormat(locale, { numeric: 'auto', style }),
  );
}

function pluralRules(locale: string): Intl.PluralRules {
  return cached(pluralRuleSets, locale, () => new Intl.PluralRules(locale));
}

function cached<T>(store: Map<string, T>, key: string, build: () => T): T {
  const existing = store.get(key);
  if (existing !== undefined) return existing;

  const built = build();
  store.set(key, built);
  return built;
}

File: crates/i18n/src/format.rs
//! The locale-aware primitives, shared; the presentation is not.
//!
//! A settings panel renders a token budget as `8,192` and a terminal status line
//! renders the same interval as `2m 30s` rather than a sentence, and that
//! difference is deliberate — a panel has a column to spend and a status line
//! does not. So what is shared here is the part that is *locale* knowledge
//! (which separator, which plural category, which unit) and not the part that
//! is *layout* judgement.
//!
//! Every function takes the locale explicitly. That is the entire fix for the
//! bug a hand-rolled comma grouper once worked around: a machine set to `de-DE`
//! rendered 8192 as `8.192`, which reads as eight in the one panel whose job is
//! making a budget legible. The problem was never that grouping is locale-aware
//! — it is that the locale was *implicit*, and so was whatever the machine
//! happened to be set to. Passed in, it is the install's locale on every machine.
//!
//! Only the primitives the terminal needs are here: number grouping, plural
//! categories, and the two pure arithmetic helpers behind "2m 30s" and "5m ago".
//! Absolute dates, compact numbers and worded intervals need CLDR pattern data
//! the terminal has no caller for, and are left to the web layer.

pub use icu_plurals::PluralCategory;
use icu_plurals::{PluralOperands, PluralRules, PluralRulesOptions};

use crate::locale::{DEFAULT_LOCALE, normalise_locale};

/// A number with the locale's grouping separator: `8,192` in `en`, `8.192` in `de`.
///
/// The separator table holds the languages this crate has been asked to render;
/// an unknown locale gets the default's separator rather than nothing, because
/// a wrong separator is legible and a missing one is a different number.
#[must_use]
pub fn format_number(value: i64, locale: &str) -> String {
    let separator = grouping_separator(locale);
    let digits = value.unsigned_abs().to_string();
    let mut grouped = String::with_capacity(digits.len() + digits.len() / 3 + 1);
    for (index, digit) in digits.chars().enumerate() {
        let remaining = digits.len() - index;
        if index > 0 && remaining.is_multiple_of(3) {
            grouped.push_str(separator);
        }
        grouped.push(digit);
    }
    if value < 0 {
        format!("-{grouped}")
    } else {
        grouped
    }
}

/// The thousands separator for a language. A table rather than a library:
/// this is the whole of what the terminal needs, and CLDR number data is a
/// dependency worth taking when a second locale ships, not before.
fn grouping_separator(locale: &str) -> &'static str {
    let normalised = normalise_locale(Some(locale));
    match normalised.split('-').next().unwrap_or("") {
        "de" => ".",
        _ => ",",
    }
}

/// Which CLDR plural category a count falls into for this locale.
///
/// The translator resolves `_one` / `_other` through the same rules, so this
/// exists for the places that build a string outside a resource — not as a
/// second pluralisation scheme.
pub fn plural_category<C: Into<PluralOperands>>(count: C, locale: &str) -> PluralCategory {
    plural_rules(locale).map_or(PluralCategory::Other, |rules| rules.category_for(count))
}

/// Cardinal plural rules for a locale, falling back to the default locale's
/// when the tag does not parse or has no data. `None` only when even the
/// default has none, which the embedded data rules out.
#[must_use]
pub fn plural_rules(locale: &str) -> Option<PluralRules> {
    let tag = normalise_locale(Some(locale));
    let parsed = icu_locale_core::Locale::try_from_str(&tag)
        .or_else(|_| icu_locale_core::Locale::try_from_str(DEFAULT_LOCALE.as_str()))
        .ok()?;
    PluralRules::try_new((&parsed).into(), PluralRulesOptions::default()).ok()
}

/// Which interval an instant falls into, without saying it in any language.
///
/// Split from the wording because the two halves belong to different owners.
/// The *thresholds* are shared — a minute is a minute in every locale, and the
/// rules about when to stop counting are product decisions rather than language
/// ones. The *wording* is not: a sidebar says `just now`, a terminal may say
/// `now`, and that is a copy decision each surface is entitled to keep.
///
/// `now_ms` is a parameter rather than a clock read so the boundaries can be
/// tested without a fake clock, and so a list rendered in one pass cannot show
/// two rows measured against two different instants.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RelativeSpan {
    /// Inside the last minute. The caller supplies its own wording.
    Now,
    /// Old enough that an interval reads worse than a date.
    Date,
    /// A whole number of units ago.
    Ago {
        /// How many of `unit`.
        value: i64,
        /// The unit a person would have picked.
        unit: SpanUnit,
    },
}

/// The units a relative interval is phrased in.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SpanUnit {
    /// Under an hour.
    Minute,
    /// Under a day.
    Hour,
    /// Under a week; past that a date reads better.
    Day,
}

/// The interval between `at_ms` and `now_ms`, in the unit a person would pick.
#[must_use]
pub fn relative_span(at_ms: i64, now_ms: i64) -> RelativeSpan {
    let elapsed = now_ms.saturating_sub(at_ms);

    // A timestamp slightly in the *future* is the present rather than a negative
    // interval: the server and the client have separate clocks, and a few
    // seconds of skew is normal rather than an error worth rendering.
    if elapsed < 60_000 {
        return RelativeSpan::Now;
    }

    let minutes = elapsed / 60_000;
    if minutes < 60 {
        return RelativeSpan::Ago {
            value: minutes,
            unit: SpanUnit::Minute,
        };
    }

    let hours = minutes / 60;
    if hours < 24 {
        return RelativeSpan::Ago {
            value: hours,
            unit: SpanUnit::Hour,
        };
    }

    let days = hours / 24;
    // Past a week the relative form stops being informative — "23d ago" is worse
    // than a date, because nobody counts back three weeks in their head.
    if days < 7 {
        RelativeSpan::Ago {
            value: days,
            unit: SpanUnit::Day,
        }
    } else {
        RelativeSpan::Date
    }
}

/// The unit a duration is led with.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DurationUnit {
    /// Below a second, where the resolution distinguishes a cache hit from a request.
    Ms,
    /// Below a minute.
    Second,
    /// Below an hour, with a seconds remainder.
    Minute,
    /// An hour or more, with a minutes remainder.
    Hour,
}

/// A duration broken into its parts, for a caller to word.
///
/// Numbers rather than a string because the surfaces disagree about the wording
/// and agree about the arithmetic — and the arithmetic is the half with the
/// boundaries that are easy to get wrong and invisible when they are: 59.9
/// seconds rounding to "60s", a sub-millisecond tool call reporting "0ms" as
/// though it never ran.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct DurationParts {
    /// The unit `value` is in.
    pub unit: DurationUnit,
    /// The leading number, already rounded for display.
    pub value: f64,
    /// The remainder in the next unit down, for `2m 30s`. Zero below a minute.
    pub remainder: i64,
    /// True below ten seconds, where one decimal is worth showing.
    pub fractional: bool,
}

/// Splits a duration in milliseconds. `None` for a negative or non-finite one.
#[must_use]
pub fn duration_parts(ms: f64) -> Option<DurationParts> {
    if !ms.is_finite() || ms < 0.0 {
        return None;
    }
    if ms < 1000.0 {
        return Some(DurationParts {
            unit: DurationUnit::Ms,
            value: ms.round(),
            remainder: 0,
            fractional: false,
        });
    }

    // Whole seconds of any duration a process could measure fit an i64.
    #[allow(clippy::cast_possible_truncation)]
    let total_seconds = (ms / 1000.0).floor() as i64;
    if total_seconds < 60 {
        // One decimal below ten seconds, where the difference is worth seeing.
        let fractional = total_seconds < 10;
        #[allow(clippy::cast_precision_loss)] // total_seconds < 60
        let whole = total_seconds as f64;
        let value = if fractional {
            (ms / 100.0).round() / 10.0
        } else {
            whole
        };
        return Some(DurationParts {
            unit: DurationUnit::Second,
            value,
            remainder: 0,
            fractional,
        });
    }

    let minutes = total_seconds / 60;
    if minutes < 60 {
        #[allow(clippy::cast_precision_loss)] // minutes < 60
        let value = minutes as f64;
        return Some(DurationParts {
            unit: DurationUnit::Minute,
            value,
            remainder: total_seconds % 60,
            fractional: false,
        });
    }

    let hours = minutes / 60;
    #[allow(clippy::cast_precision_loss)] // hours is a small count, exact in f64
    let value = hours as f64;
    Some(DurationParts {
        unit: DurationUnit::Hour,
        value,
        remainder: minutes % 60,
        fractional: false,
    })
}
